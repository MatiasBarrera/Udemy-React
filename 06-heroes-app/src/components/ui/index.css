nav {
  background-color: #082032;
  display: grid;
  grid-template-columns: repeat(7, 1fr);
  padding: 20px;
  width: 100vw;
  height: 5%;
}

.brand {
  color: #ff4c29;
  display: flex;
  font-size: 25px;
  justify-content: center;
}

.container-links {
  grid-column: 3/3;
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.container-logout {
  grid-column: 7/7;
  align-self: center;
  justify-self: center;
}

.container-logout button {
  background-color: transparent;
  border: 1px solid lightgray;
  color: white;
  cursor: pointer;
  font-weight: bolder;
  padding: 0.3rem 2rem;
}

.container-logout button:hover {
  background-color: white;
  color: black;
  transition: all 0.5s;
}

.container-user {
  grid-column: 6/7;
  align-self: center;
  justify-self: flex-end;
  color: #ff4c29;
  font-weight: bold;
}

a {
  text-decoration: none;
  color: #fff;
  font-weight: bold;
}
