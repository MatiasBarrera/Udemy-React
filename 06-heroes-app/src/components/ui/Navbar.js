import React, { useContext } from 'react'
import { Link, NavLink, useHistory } from 'react-router-dom'
import { AuthContext } from '../../auth/AuthContext'
import { types } from '../../types/types'
import '../ui/index.css'

export const Navbar = () => {
  const history = useHistory()
  const { user, dispatch } = useContext(AuthContext)

  const handleLogout = () => {
    history.replace('/login')
    dispatch({
      type: types.logout
    })
  }

  return (
    <nav>
      <Link className="brand" to="/">
        Asociaciones
      </Link>

      <div className="container-links">
        <NavLink activeClassName="active" exact to="/marvel">
          Marvel
        </NavLink>

        <NavLink activeClassName="active" exact to="/dc">
          DC Comics
        </NavLink>

        <NavLink activeClassName="active" exact to="/search">
          Search
        </NavLink>
      </div>
      <div className="container-user">{user.name}</div>
      <div className="container-logout">
        <button onClick={handleLogout}>Logout</button>
      </div>
    </nav>
  )
}
