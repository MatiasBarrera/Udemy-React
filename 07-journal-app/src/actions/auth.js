import {
  getAuth,
  signInWithPopup,
  createUserWithEmailAndPassword,
  updateProfile,
  signInWithEmailAndPassword,
  signOut
} from '@firebase/auth'
import Swal from 'sweetalert2'
import { googleAuthProvider } from '../firebase/firebase-config'
import { types } from '../types/types'
import { FinishLoading, StartLoading } from './ui'

export const startLoginEmailPassword = (email, password) => {
  return dispatch => {
    const auth = getAuth()

    dispatch(StartLoading())
    signInWithEmailAndPassword(auth, email, password)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName))
        dispatch(FinishLoading())
      })
      .catch(e => {
        Swal.fire({
          icon: 'error',
          title: 'Error',
          text: 'There is no user record corresponding to this identifier. The user may have been deleted.'
        })
        dispatch(FinishLoading())
      })
  }
}

export const startRegisterWithEmailPasswordName = (name, email, password) => {
  return dispatch => {
    const auth = getAuth()
    createUserWithEmailAndPassword(auth, email, password)
      .then(async ({ user }) => {
        await updateProfile(user, { displayName: name })

        dispatch(login(user.uid, user.displayName))
      })
      .catch(e => {
        console.log(e)
      })
  }
}

export const startGoogleLogin = () => {
  return dispatch => {
    const auth = getAuth()
    signInWithPopup(auth, googleAuthProvider).then(({ user }) => {
      dispatch(login(user.uid, user.displayName))
    })
  }
}

export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    name: displayName
  }
})

export const startLogout = () => {
  return async dispatch => {
    const auth = getAuth()
    await signOut(auth)
    dispatch(logout())
  }
}

export const logout = () => ({
  type: types.logout
})
