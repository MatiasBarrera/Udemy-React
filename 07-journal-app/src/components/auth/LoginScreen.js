import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { Link } from 'react-router-dom'
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth'
import googleLogo from '../../assets/google.logo.svg'
import { useForm } from '../../hooks/useForm'

export const LoginScreen = () => {
  const dispatch = useDispatch()
  const { loading } = useSelector(state => state.ui)

  const [formValues, handleInputChange] = useForm({
    email: 'matias@gmail.com',
    password: '123456'
  })

  const handleLogin = e => {
    e.preventDefault()
    dispatch(startLoginEmailPassword(email, password))
  }

  const handleLoginGoogle = () => {
    dispatch(startGoogleLogin())
  }

  const { email, password } = formValues
  return (
    <>
      <h3 className="login__title">Login</h3>
      <form className="auth__form" onSubmit={handleLogin}>
        <input
          type="text"
          placeholder="email"
          name="email"
          autoComplete="off"
          value={email}
          onChange={handleInputChange}
        />
        <input
          type="password"
          placeholder="password"
          name="password"
          autoComplete="off"
          value={password}
          onChange={handleInputChange}
        />
        <button className="pointer" type="submit" disabled={loading}>
          Ingresar
        </button>
        <hr />

        <div className="login__social-network">
          <p>Login with social network</p>
          <div
            className="google-btn pointer mt-1 mb-1"
            onClick={handleLoginGoogle}>
            <div className="google-icon-wrapper">
              <img
                className="google-icon"
                src={googleLogo}
                alt="google button"
              />
            </div>
            <p className="btn-text">
              <b>Sign in with google</b>
            </p>
          </div>
        </div>

        <Link to="/auth/register">Create new account</Link>
      </form>
    </>
  )
}
