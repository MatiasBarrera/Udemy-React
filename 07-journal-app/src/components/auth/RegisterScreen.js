import React from 'react'
import { Link } from 'react-router-dom'
import { useForm } from '../../hooks/useForm'
import isEmail from 'validator/lib/isEmail'
import { useDispatch, useSelector } from 'react-redux'
import { removeError, setError } from '../../actions/ui'
import { startRegisterWithEmailPasswordName } from '../../actions/auth'

export const RegisterScreen = () => {
  const dispatch = useDispatch()
  const { msgError } = useSelector(state => state.ui)

  const [formValues, handleInputChange] = useForm({
    name: 'matias',
    email: 'matias@gmail.com',
    password: '123456',
    password2: '123456'
  })

  const { name, email, password, password2 } = formValues

  const handleRegister = e => {
    e.preventDefault()

    if (isFormValid()) {
      dispatch(startRegisterWithEmailPasswordName(name, email, password))
    }
  }

  const isFormValid = () => {
    if (name.trim().length === 0) {
      dispatch(setError('Name is required'))
      return false
    }
    if (email.trim().length === 0 && !isEmail(email)) {
      dispatch(setError('Email is not valid'))
      return false
    }
    if (password !== password2 || password.length < 5) {
      dispatch(
        setError(
          'Password should be at least 6 characters and match each other '
        )
      )
      return false
    }
    dispatch(removeError())
    return true
  }

  return (
    <>
      <h3 className="login__title">Register</h3>
      <form className="auth__form" onSubmit={handleRegister}>
        {msgError && <div className="auth__alert-error">{msgError}</div>}
        <input
          type="text"
          placeholder="name"
          name="name"
          autoComplete="off"
          value={name}
          onChange={handleInputChange}
        />
        <input
          type="text"
          placeholder="email"
          name="email"
          autoComplete="off"
          value={email}
          onChange={handleInputChange}
        />
        <input
          type="password"
          placeholder="password"
          name="password"
          autoComplete="off"
          value={password}
          onChange={handleInputChange}
        />
        <input
          type="password"
          placeholder="confirm"
          name="password2"
          autoComplete="off"
          value={password2}
          onChange={handleInputChange}
        />
        <button className="pointer" type="submit">
          Register
        </button>
        <hr />

        <Link to="/auth/login">Already registered?</Link>
      </form>
    </>
  )
}
