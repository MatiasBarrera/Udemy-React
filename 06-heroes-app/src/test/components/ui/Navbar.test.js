import React from 'react'

import { Navbar } from '../../../components/ui/Navbar'
import { mount } from 'enzyme'
import { AuthContext } from '../../../auth/AuthContext'
import { MemoryRouter, Router } from 'react-router'
import { types } from '../../../types/types'

describe('Pruebas en Navbar', () => {
  const historyMock = {
    push: jest.fn(),
    replace: jest.fn(),
    location: '',
    listen: jest.fn(),
    createHref: jest.fn()
  }

  const contextValue = {
    dispatch: jest.fn(),
    user: {
      logged: true,
      name: 'Matias'
    }
  }

  const wrapper = mount(
    <MemoryRouter>
      <AuthContext.Provider value={contextValue}>
        <Router history={historyMock}>
          <Navbar />
        </Router>
      </AuthContext.Provider>
    </MemoryRouter>
  )

  test('Debe de mostrarse correctamente', () => {
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('.container-user').text().trim()).toBe('Matias')
  })

  test('Debe de llamar logout y el usar useHistory', () => {
    wrapper.find('button').simulate('click')
    expect(contextValue.dispatch).toHaveBeenCalledWith({ type: types.logout })
    expect(historyMock.replace).toHaveBeenCalledWith('/login')
  })

  afterEach(() => {
    jest.clearAllMocks()
  })
})
