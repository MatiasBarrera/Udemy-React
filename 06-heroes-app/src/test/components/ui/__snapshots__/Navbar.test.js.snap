// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en Navbar Debe de mostrarse correctamente 1`] = `
<nav>
  <a
    className="brand"
    onClick={[Function]}
  >
    Asociaciones
  </a>
  <div
    className="container-links"
  >
    <a
      aria-current={null}
      onClick={[Function]}
    >
      Marvel
    </a>
    <a
      aria-current={null}
      onClick={[Function]}
    >
      DC Comics
    </a>
    <a
      aria-current={null}
      onClick={[Function]}
    >
      Search
    </a>
  </div>
  <div
    className="container-user"
  >
    Matias
  </div>
  <div
    className="container-logout"
  >
    <button
      onClick={[Function]}
    >
      Logout
    </button>
  </div>
</nav>
`;
